substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/isolated/IsolatedSpeculationLog.java

```diff
@@ -48,19 +58,115 @@ public void collectFailedSpeculations() {
         collectFailedSpeculations0(IsolatedCompileContext.get().getClient(), handle);
     }
 
+    static class SpeculationReasonInvocationHandler implements InvocationHandler {
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return IsolatedSpeculationReasonEncoding.class.getDeclaredMethod(method.getName(), method.getParameterTypes()).invoke(args);
```

Comment:
The `invoke` call is missing a receiver of type `IsolatedSpeculationReasonEncoding`, which should first be instantiated in the supplier and passed to this invocation handler class via a constructor.

Reply:
I doubt the image build can optimize the reflective accesses, and on top of repeatedly doing lookups at runtime, we will need to register the methods of `IsolatedSpeculationReasonEncoding` as reflectively accessible in a `Feature` during the image build.
So I'd rather avoid this proxy, but I don't see how we can have a class that directly implements `SpeculationReasonEncoding` when that interface cannot be referenced in the code.

Reply:
@peter-hofer is there any way to inject this interface to `SpeculationLog` when using an openJDK which lacks it (using something like `com.oracle.svm.core.annotate.TargetClass#onlyWith`)

Alternatively, could we add the interface in the sources and compile it in only when needed? (possibly through mx).

Reply:
> The invoke call is missing a receiver of type IsolatedSpeculationReasonEncoding, which should first be instantiated in the supplier and passed to this invocation handler class via a constructor.

Shouldn't the said `IsolatedSpeculationReasonEncoding` be instantiated for each new encoding and as a result don't we need a new `SpeculationReasonInvocationHandler` for each new encoding and consequently a new proxy instance?

Reply:
> Shouldn't the said `IsolatedSpeculationReasonEncoding` be instantiated for each new encoding and as a result don't we need a new `SpeculationReasonInvocationHandler` for each new encoding and consequently a new proxy instance rendering https://github.com/oracle/graal/pull/2945/files#r519816278 impossible?

I agree, we do need a separate instance of each of these classes for each encoding operation. Not sure what that contradicts, I think you meant to link to a comment.

Reply:
> @peter-hofer is there any way to inject this interface to `SpeculationLog` when using an openJDK which lacks it (using something like `com.oracle.svm.core.annotate.TargetClass#onlyWith`)
> 
> Alternatively, could we add the interface in the sources and compile it in only when needed? (possibly through mx).

Using substitutions is a clever idea, but they don't support changes to the class hierarchy. What we could try is substituting methods of the dynamic proxy class to call the encoder directly instead. Substituting a generated class like a proxy sounds a bit fragile, though.

Reply:
> > Shouldn't the said `IsolatedSpeculationReasonEncoding` be instantiated for each new encoding and as a result don't we need a new `SpeculationReasonInvocationHandler` for each new encoding and consequently a new proxy instance rendering https://github.com/oracle/graal/pull/2945/files#r519816278 impossible?
> 
> I agree, we do need a separate instance of each of these classes for each encoding operation. Not sure what that contradicts, I think you meant to link to a comment.

It looks like after force pushing links to out-of-date discussions breaks :man_facepalming: 
New link after force pushing is https://github.com/oracle/graal/pull/2945#discussion_r519816278

Reply:
> I doubt the image build can optimize the reflective accesses, and on top of repeatedly doing lookups at runtime, we will need to register the methods of `IsolatedSpeculationReasonEncoding` as reflectively accessible in a `Feature` during the image build.
So I'd rather avoid this proxy, but I don't see how we can have a class that directly implements SpeculationReasonEncoding when that interface cannot be referenced in the code.

@peter-hofer please have a look at https://github.com/zakkak/mandrel/commit/19398016bc4bf77061268db7f5395ebd9286ebb6, it removes the proxy and relies solely on our cloned interface.
If the original `encode` gets invoked elsewhere it will essentially generate the same encoding and cache it in the corresponding `EncodedSpeculationReason` instance.
If I am not mistaken this should work without the performance penalty.

Reply:
> @peter-hofer please have a look at [zakkak@1939801](https://github.com/zakkak/mandrel/commit/19398016bc4bf77061268db7f5395ebd9286ebb6), it removes the proxy and relies solely on our cloned interface.

Looks good. I did encounter problems when the proxy class is created at image build time, so that's another reason to avoid it.
However, I would like to have two separate code paths which directly call `UnencodedSpeculationReason.class.getDeclaredField("...").get*(reason)` and `Class.forName("jdk.vm.ci.meta.EncodedSpeculationReason").getDeclaredField("...").get*(reason)` and pass the values to a method that does the encoding. That way, we can be sure that these reflective accesses will become direct field accesses during the image build.
My only worry other than that is that `EncodedSpeculationReason` may be changed in the future and this code silently becomes broken, which would have been prevented by using a proxy class.

Reply:
> My only worry other than that is that `EncodedSpeculationReason` may be changed in the future and this code silently becomes broken, which would have been prevented by using a proxy class.

Yes I also thought about this, but even with the proxy in place this would still be the case for the `UnencodedSpeculationReason` path, no?
I guess we should implement some kind of test and add it to `mx gate` so that if/when `EncodedSpeculationReason` changes we get notified (the simplest I can think of is comparing the result of the copied methods with the result of the corresponding methods in `EncodedSpeculationReason`).

Reply:
@peter-hofer the test is now added to the PR, I hope this is enough to detect potential future changes in `EncodedSpeculationReason` that could cause issues.